checker-util/src/main/java/org/checkerframework/checker/regex/util/RegexUtil.java

```diff
@@ -362,57 +398,128 @@ private static int getGroupCount(Pattern p) {
     }
 
     /**
-     * Return the non-null groups in the argument.
+     * Returns a list of groups other than 0, that are guaranteed to be non-null given that the
+     * regular expression matches the String.
      *
-     * @param p pattern to be analysed
+     * @param regexp pattern to be analysed
+     * @param n number of capturing groups in the pattern
      * @return the groups that are non-null in the argument
      */
-    private static List<Integer> getNonNullGroups(Pattern p) {
-        String regexp = p.pattern();
-        int n = getGroupCount(p);
+    public static List<Integer> getNonNullGroups(String regexp, int n) {
         List<Integer> nonNullGroups = new ArrayList<>();
-        for (int i = 0; i <= n; i++) {
+        // Initialize the list with all elements; the below code will remove some.
+        for (int i = 1; i <= n; i++) {
             nonNullGroups.add(i);
         }
+
         ArrayDeque<Integer> openingIndices = new ArrayDeque<>();
+        ArrayDeque<Boolean> openingWasGroup = new ArrayDeque<>();
+        int insideList = 0; // The part being traversed now is inside [...].
+        boolean quoted = false; // The part being traversed now is inside \Q ... \E.
+        boolean escaped = false; // The character just before the current one was '\'.
         int group = 0;
-        boolean squareBracketOpen = false;
-        boolean escaped = false;
-        boolean notCapturing = false;
-        for (int i = 0; i < regexp.length(); i++) {
-            if (!escaped && !squareBracketOpen && regexp.charAt(i) == '(') {
-                if (i != regexp.length() - 1 && regexp.charAt(i + 1) == '?') {
-                    notCapturing = true;
-                    continue;
+
+        /**
+         * If you encounter '(' and it is not inside a list or a quote and is not preceded by a '\',
+         * then it is a special group. If it is followed by a '?', it will be either a named
+         * capturing group or some other special construct. A named capturing group has a '<'
+         * followed by the '?'. The boolean stack is to figure out whether a ')' closes a capturing
+         * group or a special construct.
+         *
+         * <p>If you encounter a '\' and it is not inside a quote, it defines some sort of flag or
+         * special character. If you are inside a quote and '\' is followed immediately by 'E', it
+         * is there to mark the end of the quote so '\' is not there in a literal sense and can be
+         * considered as escaped.
```

Comment:
What does "considered as escaped" mean?

Reply:
It means that '\' is not in a literal sense inside a quote. It is there as a part of an escape sequence.

Reply:
OK.  Please clarify the documentation, because readers won't know what you mean by "considered as escaped".